### PR TITLE
fix(entry): added battery level check

### PR DESCRIPTION
**Pull request by ThornWalli**
- occurs when battery is not charging and <= 20%.
- added message in speedkitLayer.

#786

### CODE REVIEW

**Review by StephanGerbeth — CHANGES_REQUESTED** [review]

File: src/runtime/utils/entry.mjs
[the remark this answers is not in the document]
Comment (ThornWalli) [reply src/runtime/utils/entry.mjs:1]: There are two different checks.

1. check battery api
2. check if a video can be played without user interaction

https://github.com/GrabarzUndPartner/nuxt-speedkit/blob/2b9206cc1f999ef474041fc237c7e303e633ed5b/src/runtime/utils/entry.mjs#L65-L112

Checking sequence

https://github.com/GrabarzUndPartner/nuxt-speedkit/blob/2b9206cc1f999ef474041fc237c7e303e633ed5b/src/runtime/utils/entry.mjs#L59C1-L64

The check is placed in the entry before the performance check and triggers a separate message in the layer.

https://github.com/GrabarzUndPartner/nuxt-speedkit/blob/2b9206cc1f999ef474041fc237c7e303e633ed5b/src/runtime/tmpl/entry.mjs#L46-L60

File: src/module.mjs
[the remark this answers is not in the document]
Comment (ThornWalli) [reply src/module.mjs:1]: ready!

https://github.com/GrabarzUndPartner/nuxt-speedkit/blob/54ceb684812d31f777ec67e8bbb5ca841854d833/src/module.mjs#L147-L150

File: src/runtime/utils/entry.mjs [hunk 1]
```diff
@@ -56,3 +56,57 @@ export function initReducedView() {
...
+async function isBatteryLow() {
+  const MIN_BATTERY_LEVEL = 0.2;
+
+  try {
+    const battery = await window.navigator.getBattery();
+    if (!battery.charging && battery.level <= MIN_BATTERY_LEVEL) {
+      throw new Error('Battery is low.');
+    }
+  } catch (error) {
+    // Ignore Check
```
Comment (StephanGerbeth) [inline src/runtime/utils/entry.mjs:82]: silent catch?

File: src/runtime/utils/entry.mjs [hunk 2]
```diff
@@ -56,3 +56,57 @@ export function initReducedView() {
...
+ * Checking whether a video can be played.
+ * This check is for IOS and checks if the power saving mode is enabled.
+ *
+ * In this case no video will be played automatically.
+ */
+export async function canVideoPlay(blob) {
+  const VIDEO_TIMEOUT = 250;
+
+  const video = document.createElement('video');
+  video.muted = true;
```
Comment (StephanGerbeth) [inline src/runtime/utils/entry.mjs:96]: do we need that. it seems to be a duplicate -> next line

File: src/runtime/utils/entry.mjs [hunk 3]
```diff
@@ -56,3 +56,57 @@ export function initReducedView() {
...
+  video.setAttribute('muted', 'muted');
+  video.setAttribute('playsinline', 'playsinline');
+  video.src = URL.createObjectURL(blob);
+
+  const timeout = setTimeout(() => {
+    throw new Error('Video playback not possible. Reason: timeout');
+  }, VIDEO_TIMEOUT);
+
+  try {
+    await video.play();
```
Comment (StephanGerbeth) [inline src/runtime/utils/entry.mjs:106]: explain me, what should happen here in combination with the timeout.
i think we should wait for the playing event? and perhaps there is another event we can use to detect it is not playing cause of weak battery

**Review by StephanGerbeth — CHANGES_REQUESTED** [review]

File: src/runtime/utils/entry.mjs (hunk 3 above)
Comment (ThornWalli) [reply src/runtime/utils/entry.mjs:94]: Video timeout was removed.

Now only the reject from the video `play()` is caught.

https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/play

File: src/runtime/utils/entry.mjs (hunk 2 above)
Comment (ThornWalli) [reply src/runtime/utils/entry.mjs:89]: All `setAttribute` have been removed and replaced with set properties.

File: src/runtime/utils/entry.mjs (hunk 1 above)
Comment (ThornWalli) [reply src/runtime/utils/entry.mjs]: The unnecessary catch has been removed.

In the entry now `console.warn` are built in, these show the respective case as a warning.

image

File: src/runtime/utils/entry.mjs [hunk 4]
```diff
@@ -90,23 +85,14 @@ async function isBatteryLow() {
...
-  const timeout = setTimeout(() => {
-    throw new Error('Video playback not possible. Reason: timeout');
-  }, VIDEO_TIMEOUT);
-
   try {
     await video.play();
-    clearTimeout(timeout);
   } catch (error) {
-    clearTimeout(timeout);
     throw new Error("Video playback not possible. Reason: can't play video");
```
Comment (StephanGerbeth) [inline src/runtime/utils/entry.mjs:96]: please consign a meaningful error message

**Review by StephanGerbeth — CHANGES_REQUESTED** [review]

File: src/runtime/utils/entry.mjs
[the remark this answers is not in the document]
Comment (StephanGerbeth) [reply src/runtime/utils/entry.mjs:1]: Please use the silent catch to switch to video detection.

File: src/media/video.mp4
Comment (StephanGerbeth) [inline src/media/video.mp4:1]: do we use an additional file? Or can we also use a super small base64 string?

File: src/runtime/utils/entry.mjs
[the remark this answers is not in the document]
Comment (StephanGerbeth) [reply src/runtime/utils/entry.mjs:1]: Please use the silent catch to switch to video detection.

**Review by StephanGerbeth — CHANGES_REQUESTED** [review]

**Comment by Object905** [comment]
Any news on this? Really like to see this in v3-next.

**Comment by ThornWalli** [comment]
Hello @Object905 ,

will be in shortly 😉

Just planning the renaming of the package there it must be included 🙂
https://github.com/GrabarzUndPartner/nuxt-speedkit/issues/921

But will probably be available before that.

**Review by StephanGerbeth — APPROVED** [review]

**Pull request merged**